File: solutions/sonnet-1/sonnet/javascript/dijkstra.js
const lines = require('fs').readFileSync('/dev/stdin', 'utf8').split('\n');

let idx = 0;
const [n, m] = lines[idx++].trim().split(/\s+/).map(Number);

const adj = Array.from({ length: n }, () => []);
for (let i = 0; i < m; i++) {
  const [u, v, w] = lines[idx++].trim().split(/\s+/).map(Number);
  adj[u].push([v, w]);
  adj[v].push([u, w]);
}

const [s, t] = lines[idx++].trim().split(/\s+/).map(Number);

if (s === t) {
  console.log(0);
} else {
  const dist = new Array(n).fill(Infinity);
  dist[s] = 0;
  const visited = new Array(n).fill(false);

  for (let iter = 0; iter < n; iter++) {
    let u = -1;
    let best = Infinity;
    for (let i = 0; i < n; i++) {
      if (!visited[i] && dist[i] < best) {
        best = dist[i];
        u = i;
      }
    }
    if (u === -1) break;
    visited[u] = true;
    for (const [v, w] of adj[u]) {
      if (dist[u] + w < dist[v]) {
        dist[v] = dist[u] + w;
      }
    }
  }

  console.log(dist[t] === Infinity ? -1 : dist[t]);
}

File: solutions/sonnet-1/sonnet/ruby/dijkstra.rb
require 'set'

def main
  input = $stdin.read.split("\n").reject { |l| l.strip.empty? }
  idx = 0
  n, m = input[idx].split.map(&:to_i)
  idx += 1

  adj = Array.new(n) { [] }
  m.times do
    u, v, w = input[idx].split.map(&:to_i)
    idx += 1
    adj[u] << [v, w]
    adj[v] << [u, w]
  end

  s, t = input[idx].split.map(&:to_i)

  if s == t
    puts 0
    return
  end

  dist = Array.new(n, Float::INFINITY)
  dist[s] = 0
  visited = Array.new(n, false)

  n.times do
    u = -1
    best = Float::INFINITY
    n.times do |i|
      if !visited[i] && dist[i] < best
        best = dist[i]
        u = i
      end
    end
    break if u == -1

    visited[u] = true

    adj[u].each do |v, w|
      if dist[u] + w < dist[v]
        dist[v] = dist[u] + w
      end
    end
  end

  if dist[t] == Float::INFINITY
    puts(-1)
  else
    puts dist[t].to_i
  end
end

main
